### PR TITLE
Authentication (and Authorization)

### PR DESCRIPTION
**Pull request by Zanibas**
Implements:

- Refresh and access tokens
- Stores those tokens in cookies and also provides them in the mutation response.
- Login and register pages now work at an MVP level, without redirecting

TODO:
- Cookies need to actually save when not accessing via port 4000, instead from port 3000
- Authorization checks for protected pages
- Need to test if the viewer query works
- Method for invalidating tokens
- Probably some more stuff?

### CODE REVIEW

File: clients/web/src/graphql/queries/ProjectOwner.gql [hunk 1]
```diff
@@ -3,3 +3,36 @@ query ProjectOwner($id: String!) {
...
+  }
+}
+
+mutation Login($login: String!, $password: String!, $remember: Boolean!) {
+  login(login: $login, password: $password, remember: $remember) {
+    userId
+    refreshToken
+    accessToken
+  }
+}
```
Comment (jsbroks) [inline clients/web/src/graphql/queries/ProjectOwner.gql:27]: Each query/mutation should be in there own file. and mutations should be in the mutations dir

File: clients/web/src/pages/login.tsx [hunk 2]
```diff
@@ -13,38 +13,61 @@ import {
...
+      })
+      console.log(result.data)
+      setSuccess(
+        `Login succeeded for ${result.data.login.userId}. Redirecting.`
+      )
+    } catch (e) {
+      setSuccess('')
+      setError('Invalid credentials.')
+    }
+  }
```
Comment (jsbroks) [inline clients/web/src/pages/login.tsx:54]: Remove console log and  `${result.data.login.userId}` username would probably be better

File: clients/web/src/pages/register.tsx [hunk 3]
```diff
@@ -12,22 +12,45 @@ import {
...
+      })
+      setError('')
+      setSuccess(
+        `Registration succeeded. Your user id is: ${result.data.register?.userId}`
+      )
+    } catch (e) {
+      setSuccess('')
+      setError('Registration failed.')
+    }
+  }
```
Comment (jsbroks) [inline clients/web/src/pages/register.tsx:53]: remove `eslint-disable-next-line no-console` + using username

File: services/api/src/context.ts [hunk 4]
```diff
@@ -1,16 +1,19 @@
...
+type FastifyResponse = { request: FastifyDecoratedRequest; reply: FastifyReply }
 const db = new PrismaClient()
 export interface Context extends ApolloContext {
   db: PrismaClient
   reply: FastifyReply
-  request: FastifyRequest
+  request: FastifyDecoratedRequest
 }
 
+export const TOKEN_SECRET = 'REPLACETHISWITHSOMETHINGELSE'
```
Comment (jsbroks) [inline services/api/src/context.ts:15]: use `process.env.TOKEN_SECRET ?? 'default'`

File: services/api/src/modules/ProjectOwner.ts [hunk 5]
```diff
@@ -1,77 +1,187 @@
...
-      }
-    })
+    t.id(PO.id.name, PO.id)
+    t.string(PO.name.name, PO.name)
+    t.dateTime(PO.createdAt.name, PO.createdAt)
+    t.dateTime(PO.updatedAt.name, PO.updatedAt)
+    t.dateTime(PO.lastSeenAt.name, PO.lastSeenAt)
+    t.boolean(PO.disabled.name, PO.disabled)
+    t.field(PO.type.name, PO.type)
+    t.field(PO.role.name, PO.role)
```
Comment (jsbroks) [inline services/api/src/modules/ProjectOwner.ts:31]: I would use full name `ProjectOwner` to be explict

File: services/api/src/modules/ProjectOwner.ts [hunk 6]
```diff
@@ -1,77 +1,187 @@
...
+    t.string(P.company.name, P.company)
+    t.string(P.avatarUrl.name, P.avatarUrl)
+    t.string(P.websiteUrl.name, P.websiteUrl)
+    t.string(P.description.name, P.description)
+    t.string(P.location.name, P.location)
+    t.string(P.githubId.name, P.githubId)
+    t.string(P.facebookId.name, P.facebookId)
+    t.string(P.twitterId.name, P.twitterId)
+  }
+})
```
Comment (jsbroks) [inline services/api/src/modules/ProjectOwner.ts:50]: Full name for Project

File: services/api/src/modules/ProjectOwner.ts [hunk 7]
```diff
@@ -1,77 +1,187 @@
...
-      },
-      resolve(_root, args, ctx) {
-        return ctx.db.projectOwner.create({
-          data: args
-        })
+    t.string('accessToken')
+    t.string('refreshToken')
+    t.string('userId')
+  }
+})
```
Comment (jsbroks) [inline services/api/src/modules/ProjectOwner.ts:79]: I think the auth payload should have a user resolver, which can return the user Id

File: services/api/src/tokens.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
+import { FastifyReply } from 'fastify'
+import { FastifyDecoratedRequest } from './context'
+import { TOKEN_SECRET } from './context'
+import { sign, verify } from 'jsonwebtoken'
+import { addDays, addHours, addMinutes } from 'date-fns'
+import { Role, PrismaClient } from '@prisma/client'
+
+export enum LoginOrRegister {
```
Comment (jsbroks) [inline services/api/src/tokens.ts:8]: Maybe `AuthAction` would be a bettername

File: services/api/src/tokens.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+  const refreshToken = createToken(refreshData, expiresIn)
+  const accessToken = createToken(accessData, '15min')
+
+  setToken(reply, 'refresh-token', refreshToken, expires)
+  setToken(reply, 'access-token', accessToken, addMinutes(new Date(), 15))
+  return { refreshToken, accessToken, userId }
+}
+
+// hook function to run before accessing pages that require auth.
```
Comment (jsbroks) [inline services/api/src/tokens.ts:89]: Use JSDocs format

File: shared/prisma/schema.vuerd.json [hunk 10]
```diff
@@ -0,0 +1,346 @@
+{
```
Comment (jsbroks) [inline shared/prisma/schema.vuerd.json:1]: What is this file?

**Review by jsbroks — CHANGES_REQUESTED** [review]

**Comment by jsbroks** [comment]
You can set the PR to WIP

File: services/api/src/tokens.ts (hunk 9 above)
Comment (jsbroks) [reply services/api/src/tokens.ts:89]: Is there not a `FastifyHookFunction` type? Or something like it?

File: services/api/src/tokens.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  const signedRefreshToken = request.cookies['refresh-token']
+  const signedAccessToken = request.cookies['access-token']
+
+  // if we don't find any, continue onwards
+  if (!signedRefreshToken && !signedAccessToken) return
+
+  // if we have a valid access token, use the data inside.
+  try {
+    // check if the signature is valid
+    const unsignedAccessToken = request.unsignCookie(signedAccessToken)
```
Comment (jsbroks) [inline services/api/src/tokens.ts:104]: `const {valid, value} = request.unsignCookie(signedAccessToken)`

File: services/api/src/tokens.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+      login: data.login,
+      count: data.count,
+      remember: data.remember
+    }
+
+    // reissue tokens
+    issueTokens(reply, userData, LoginOrRegister.LOGIN)
+    request.user = userData
+  } catch {
+    return
```
Comment (jsbroks) [inline services/api/src/tokens.ts:144]: not needed

File: clients/web/src/pages/login.tsx (hunk 2 above)
Comment (Zanibas) [reply clients/web/src/pages/login.tsx:54]: This is just a placeholder for now to confirm success, will redirect to home page of user after login.

**Pull request merged**